include/envoy/tcp/upstream.h

```diff
@@ -0,0 +1,132 @@
+   */
+  virtual void onGenericPoolReady(StreamInfo::StreamInfo* info,
+                                  std::unique_ptr<GenericUpstream>&& upstream,
+                                  Upstream::HostDescriptionConstSharedPtr& host,
+                                  const Network::Address::InstanceConstSharedPtr& local_address,
+                                  Ssl::ConnectionInfoConstSharedPtr ssl_info) PURE;
+
+  /**
+   * Called to indicate a failure for GenericConnPool::newStream to establish a stream.
+   *
+   * @param reason supplies the failure reason.
+   * @param host supplies the description of the host that caused the failure. This may be nullptr
+   *             if no host was involved in the failure (for example overflow).
+   */
+  virtual void onGenericPoolFailure(ConnectionPool::PoolFailureReason reason,
+                                    Upstream::HostDescriptionConstSharedPtr host) PURE;
+};
+
+// Interface for a generic Upstream, which can communicate with a TCP or HTTP
+// upstream.
+class GenericUpstream {
+public:
+  virtual ~GenericUpstream() = default;
+
+  /**
+   * Enable/disable further data from this stream.
+   *
+   * @param disable true if the stream should be read disabled, false otherwise.
+   * @return returns true if the disable is performed, false otherwise
+             (e.g. if the connection is closed)
+   */
+  virtual bool readDisable(bool disable) PURE;
+
+  /**
+   * Encodes data upstream.
+   * @param data supplies the data to encode. The data may be moved by the encoder.
+   * @param end_stream supplies whether this is the last data to encode.
+   */
+  virtual void encodeData(Buffer::Instance& data, bool end_stream) PURE;
+
+  /**
+   * Adds a callback to be called when the data is sent to the kernel.
+   * @param cb supplies the callback to be called
+   */
+  virtual void addBytesSentCallback(Network::Connection::BytesSentCb cb) PURE;
+
+  /**
+   * Called when an event is received on the downstream connection
+   * @param event supplies the event which occurred.
+   * @return the underlying ConnectionData if the event is not "Connected" and draining
+             is supported for this upstream.
+   */
+  virtual Tcp::ConnectionPool::ConnectionData*
+  onDownstreamEvent(Network::ConnectionEvent event) PURE;
+};
+
+using GenericConnPoolPtr = std::unique_ptr<GenericConnPool>;
+
+/*
+ * A factory for creating generic connection pools.
+ */
+class GenericConnPoolFactory : public Envoy::Config::TypedFactory {
+public:
+  ~GenericConnPoolFactory() override = default;
+
+  /*
+   * @param cluster_name the name of the cluster to use
+   * @param cm the cluster manager to get the connection pool from
+   * @param hostname the hostname to use if doing connect tunneling, absl::nullopt otherwise.
+   * @param context the load balancing context for this connection.
+   * @param upstream_callbacks the callbacks to provide to the connection if succesfully created.
```

Comment:
Spell check error from CI:

-   * @param upstream_callbacks the callbacks to provide to the connection if succesfully created.
+   * @param upstream_callbacks the callbacks to provide to the connection if successfully created.